### ISSUE TITLE
Performance of `torch.compile` is significantly slowed down under `torch.inference_mode`

### ISSUE THREAD

**Issue by xmfbit** [pytorch/pytorch#114119]
### 🐛 Describe the bug

I have observed that when using `torch.compile` to optimize a model, the performance significantly degrades during inference under `torch.inference_mode` . In fact, it is even worse than the performance of the non-optimized model. However, when I place the compilation process within the context of `torch.inference_mode` , the performance issues are resolved.

I have posted an issue on PyTorch forum ([this thread](https://discuss.pytorch.org/t/performance-of-torch-compile-is-significantly-slowed-down-under-torch-inference-mode/191939/2)) and is advised to post it here

Here is the code to reproduce my result

```
import torch
from torchvision.models import resnet18
import sys
from contextlib import nullcontext

# enable inference mode when compile or benchmark?
enable_when_compile = sys.argv[1] == "true"
enable_when_benchmark = sys.argv[2] == "true"


def _benchmark(
    iters,
    f,
    context,   # with torch.inference or not
    *args,
    **kwargs,
) -> float:
    """Estimates the average time duration for a single inference call in second
    Returns:
        estimated average time duration in second for a single inference call
    """
    with context():
        f(*args, **kwargs)
    torch.cuda.synchronize()
    start_event = torch.cuda.Event(enable_timing=True)
    end_event = torch.cuda.Event(enable_timing=True)
    with context():
        start_event.record()
        for _ in range(iters):
            f(*args, **kwargs)
        end_event.record()
    torch.cuda.synchronize()
    elapsed_time_s = start_event.elapsed_time(end_event) * 1.0e-3
    avg_time_s = elapsed_time_s / iters
    print("Estimated average time duration: {:.6f} s".format(avg_time_s))
    return avg_time_s


class BenchmarkRunner(object):
    def __init__(self, use_inference_mode: bool):
        self.context = nullcontext if not use_inference_mode else torch.inference_mode

    def __call__(self, iters, f, *args, **kwargs) -> float:
        return _benchmark(iters, f, self.context, *args, **kwargs)


@torch.no_grad()
def run():
    input = [torch.rand(8, 3, 224, 224).to(torch.device("cuda"), dtype=torch.float16)]
    net = resnet18(pretrained=False).cuda().half()
    net.eval()
    context = nullcontext if not enable_when_compile else torch.inference_mode
    compiled = torch.compile(net, mode="reduce-overhead", backend="inductor")
    with context():
        _ = compiled(*input)

    latency_compiled = BenchmarkRunner(enable_when_benchmark)(10, compiled, *input)
    latency_torch = BenchmarkRunner(enable_when_benchmark)(10, net, *input)

    print(f"enable inference mode when compile: {enable_when_compile}")
    print(f"enable inference mode when benchmark: {enable_when_benchmark}")
    print(f"not compiled latency: {latency_torch}")
    print(f"compiled latency: {latency_compiled}, speed up: {latency_torch / latency_compiled}")

if __name__ == "__main__":
    run()
```

### Error logs

My resultI tested in `torch==2.1.0+cu118` on an A10 GPU:

```
# 1. both are disabled, speed up 2.5X, it's great!
enable inference mode when compile: False
enable inference mode when benchmark: False
not compiled latency: 0.002047449684143066
compiled latency: 0.0008150367736816407, speed up: 2.5120948529652662

# 2.  both are enabled, speed up 2.35X, great work!
enable inference mode when compile: True
enable inference mode when benchmark: True
not compiled latency: 0.0019193056106567383
compiled latency: 0.0008156064033508301, speed up: 2.3532252846121366

# 3. enable torch.inference_mode when benchmark, disabled when compiling
# the compiled model is slower than naive torch model ! help me!
# and the naive model's latency is still ~0.002
enable inference mode when compile: False
enable inference mode when benchmark: True
not compiled latency: 0.0019556352615356445
compiled latency: 0.010692658996582031, speed up: 0.18289513040309005

# 4. disable in benchmark and enable in compiling, the result is also disappointed
enable inference mode when compile: True
enable inference mode when benchmark: False
not compiled latency: 0.0020351423263549806
compiled latency: 0.011507730865478516, speed up: 0.17685001067066186
```

### Minified repro

This issue is not related to compile error, so I skip this.

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.31

Python version: 3.10.12 (main, Jul  5 2023, 18:54:27) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.143-2-velinux1-amd64-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A10
Nvidia driver version: 470.129.06
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 57 bits virtual
CPU(s):                          112
On-line CPU(s) list:             0-39
Off-line CPU(s) list:            40-111
Thread(s) per core:              0
Core(s) per socket:              28
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           106
Model name:                      Intel(R) Xeon(R) Platinum 8336C CPU @ 2.30GHz
Stepping:                        6
CPU MHz:                         2300.000
BogoMIPS:                        4600.00
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       1.3 MiB
L1i cache:                       896 KiB
L2 cache:                        35 MiB
NUMA node0 CPU(s):               0-55
NUMA node1 CPU(s):               56-111
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves wbnoinvd arat avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid md_clear arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.26.2
[pip3] onnx==1.13.1
[pip3] onnx-graphsurgeon==0.3.26
[pip3] onnxruntime==1.14.1
[pip3] optree==0.10.0
[pip3] pytorch-triton==2.1.0+6e4932cda8
[pip3] torch==2.1.0+cu118
[pip3] torch-tensorrt==1.4.0
[pip3] torchaudio==2.1.0+cu118
[pip3] torchvision==0.16.0+cu118
[pip3] triton==2.1.0
[conda] magma-cuda118             2.6.1                         1    pytorch
[conda] numpy                     1.26.2                   pypi_0    pypi
[conda] optree                    0.10.0                   pypi_0    pypi
[conda] pytorch-triton            2.1.0+6e4932cda8          pypi_0    pypi
[conda] torch                     2.1.0+cu118              pypi_0    pypi
[conda] torch-tensorrt            1.4.0                    pypi_0    pypi
[conda] torchaudio                2.1.0+cu118              pypi_0    pypi
[conda] torchvision               0.16.0+cu118             pypi_0    pypi
[conda] triton                    2.1.0                    pypi_0    pypi
```

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519
```[tasklist]
### Tasks
```

**Comment by ezyang** [comment]
The logs suggest so too: I see two compiles when I do `TORCH_LOGS=+dynamo`

**Comment by xmfbit** [comment]
@ezyang The warm up is here:
![image](https://github.com/pytorch/pytorch/assets/6002490/3e01c5db-2236-464a-b587-209499d6159b)

**Comment by masnesral** [comment]
@xmfbit I think this recent document from ezyang explains it: https://docs.google.com/document/d/1y5CRfMLdwEoF1nTk9q8qEu1mgMUuUtvhklPKJ2emLU8/edit#heading=h.z4ib9n4155bj

If I add a _2nd_ warmup run to your repro, then I see the expected performance. I'll close this issue, but feel free to reopen if you don't think your issue is resolved.
